full-stack-tests/core/webpack.config.js

```diff
@@ -35,7 +35,7 @@ function createConfig(shouldInstrument) {
       devtoolModuleFilenameTemplate: "file:///[absolute-resource-path]"
     },
     devtool: "nosources-source-map",
-    resolve: { mainFields: ["main", "module"] },
+    // resolve: { mainFields: ["main", "module"] },
```

Comment:
what happens when w/ storage-blob, if you set this to just 
```suggestion
    resolve: { mainFields: ["main"] },
```
telling webpack to only use cjs?

Reply:
I think the goal should be to only use ESM in this case, because we need the browser-specific exports.

Reply:
I just tried playing around with it and changed it to `["module", "main"]` - now I don't have any webpack errors, I'll see if I get any other problems with this

Reply:
so thats actually the default, and you dont need to explicitly specify it like this. im surprised you didnt run into any build issues, there was a reason we had to change the order/preference around last time, just don't recall what off the top of my head

Reply:
well if that's the default, I was probably just hitting some weird cache issue because I had errors with this commented out. Thanks for getting me back on track

Reply:
i now recall why we had to prefer cjs over esm; some tests are usign internal apis that arent exported from the main barrel. so we have to import like { xyz } from itwin/core-frontend/cjs/webgl/...

If we have webpack prefer esm, then sometimes you will get multiple IModelApp instances (one for cjs and for esm)

Reply:
I hate webpack 😠

Reply:
i ran into the same error youre seeing in the pipeline here in my webpack 5 pr; until we expose the internal frontend apis in the main barrel, we'll need to continue to use cjs in the webpacked bundle.

looks like object-storage-core only outputs cjs anyway. should be ok to use the cjs ver of  azure/storage-blob for the frontend here (i hope)

Reply:
The problem is that `@azure/storage-blob` doesn't have a frontend-specific cjs barrel. It has frontend overrides for esm, so you need to import using esm to avoid getting all the nodejs-specific imports (`fs`, `os` etc.)

Reply:
I don't think it matters here that the package doesn't have frontend specific cjs barrel, it does have a cjs barrel which I guess uses node internals like you said. But if that's the case we can just add an option to fallback to a polyfills. Not ideal in production in an actual use case but since this is just a test I think it's ok. I have an example of providing fallbacks in my PR.

Reply:
since the dependency is in core-frontend, we'll have to come up with a config we can use for real frontends too.

Reply:
I just filed this, I hope it gets picked up but it might take a while... https://github.com/Azure/azure-sdk-for-js/issues/22747

Reply:
@aruniverse I just pushed a silly workaround, it seems to work fine without breaking our existing webpack config. It's not ideal because I had to hardcode some module paths and I don't think those are technically part of the public API, so they might change later. I'm still looking for a way to force webpack to use the `module` field for these two packages - that should have the same effect without me having to hardcode any internal file paths.

As for lazy loading these Azure packages, I've been thinking of adding a Javascript shim (with a matching *.d.ts file) that would basically just contain `await import(/* webpackChunkName: "object-storage" */ "@itwin/object-storage-azure/lib/frontend")` and wouldn't be recompiled to `require` by Typescript. I still don't have a clue where this shim could be located, though. Any thoughts? Better ideas?

Reply:
I'm ok with both hacks here, but i think we should get @wgoehrig 's take

Reply:
@wgoehrig pretty please

Reply:
Sorry I'm so late responding... 😞 

After digging into this, I have two thoughts on how I'd prefer to handle this:

- I'm also OK with the hacks here - seems like a workable short-term solution.
- For a more permanent solution, I'm wondering if we can't essentially reverse our CJS/ESM preferences across the repo, _without_ dropping support for CJS.  This would entail:
    - Switching the behavior of `rush build` (along w/ corresponding npm scripts) so that `build` would only build ESM, while `build:ci` would continue do both
      - We may need to then always build both for a few common packages, as they'll still only be consumed via CJS in backend unit tests, but the big browser-only packages (like core-frontend) would still only build once.  That could slightly slow down `rush build` though...
      - This would still have to be done via some tsc CLI args in npm scripts (as we do today) - I don't think we can safely change the default module system in tsconfig-base.json in 3.x, since that would be a breaking change to other consumers of that config.
    - Updating all our ESM-supporting packages to specify their ESM barrel in their package.json `"typings":` fields (instead of CJS)
      - This may break downstream consumers that import from outside of barrel files (i.e., `.../lib/cjs/...`), but that probably means they were already using internal APIs or were likely to hit run-time or bundling issues anyway.  From a quick search, it looks like we've been pretty good about this within the repo at least, only using said imports in a handful of these "fullstack" tests...
    - Updating all "fullstack-tests" and "test-apps" packages to build ESM instead of CJS
      - This means removing both this hack here as well as reverting back to webpack's default behavior for `mainFields`.
      - Also means changing all explicit/private (`.../lib/cjs/...`) imports in our tests to explicitly import ESM instead  (`.../lib/esm/...`).

  Not only would that help us clean up/simplify these webpack configs, but it would also make it easier/faster for folks to build our test-apps (specifically DTA) without accidentally incurring CJS-related performance issues.

  FWIW, I took a quick pass at this locally and confirmed it would at least be possible to run the core and rpc fullstack-tests this way.  Of course, that's not to say there won't be some big gotchas here... The biggest downside I can foresee at the moment would be that with these changes, you'd always have to do a **`rush build:ci`** before running UI and presentation tests (since those still need to run in Node and thus will be stuck using CJS).  Maybe that's an acceptable trade-off to make the rest of our tests more consistent with what consumers of these packages will actually be running/deploying though?

Reply:
@wgoehrig I agree with the two approaches here for what to do going forward. I just had a conversation this morning with @GerardasB about a consumer that accidentally included both an esm and cjs module and had an issue with two separate contexts.

Switching to set the `"typings"` seems like a good way to help "encourage" people to do the right thing and not use internal APIs or the direct imports.

Reply:
@wgoehrig you were right, the webpack config we have in react-scripts worked! I just tested it with a local itwin viewer and it worked. I'm a bit hesitant to merge this just before my vacation, because even though I've tested it, if something doesn't work right I won't be here to fix it. Feel free to merge it while I'm out if you feel like it.